modules/apps/commerce/commerce-shipping-engine/src/mail/java/com/liferay/commerce/shipping/engine/internal/FunctionCommerceShippingEngine.java

```diff
@@ -0,0 +1,335 @@
+
+			JSONObject jsonObject = _jsonFactory.createJSONObject(
+				new String(
+					_portalCatapult.launch(
+						commerceOrder.getCompanyId(), Http.Method.POST,
+						_functionCommerceShippingEngineConfiguration.
+							oAuth2ApplicationExternalReferenceCode(),
+						_getPayloadJSONObject(commerceContext, commerceOrder),
+						_functionCommerceShippingEngineConfiguration.
+							enabledShippingEngineOptionsPath(),
+						currentUser.getUserId()
+					).get()));
+
+			return _getCommerceShippingOptions(jsonObject);
+		}
+		catch (Exception exception) {
+			_log.error(exception);
+
+			throw new CommerceShippingEngineException(exception.getMessage());
+		}
+	}
+
+	@Override
+	public String getKey() {
+		return _functionCommerceShippingEngineConfiguration.key();
+	}
+
+	@Override
+	public String getName(Locale locale) {
+		try {
+			User currentUser = _userService.getCurrentUser();
+
+			JSONObject jsonObject = _jsonFactory.createJSONObject(
+				new String(
+					_portalCatapult.launch(
+						currentUser.getCompanyId(), Http.Method.POST,
+						_functionCommerceShippingEngineConfiguration.
+							oAuth2ApplicationExternalReferenceCode(),
+						JSONUtil.put("locale", locale),
+						_functionCommerceShippingEngineConfiguration.
+							shippingEngineLocalizedNamePath(),
+						currentUser.getUserId()
+					).get()));
+
+			return jsonObject.getString("name");
+		}
+		catch (Exception exception) {
+			_log.error(exception);
+		}
+
+		return StringPool.BLANK;
+	}
+
+	@Activate
+	protected void activate(Map<String, Object> properties) {
+		_functionCommerceShippingEngineConfiguration =
+			ConfigurableUtil.createConfigurable(
+				FunctionCommerceShippingEngineConfiguration.class, properties);
+	}
+
+	@Deactivate
+	protected void deactivate() throws PortalException {
+		String key = getKey();
+
+		if (key == null) {
+			return;
+		}
+
+		List<CommerceShippingMethod> commerceShippingMethods =
+			_commerceShippingMethodLocalService.getCommerceShippingMethods(
+				QueryUtil.ALL_POS, QueryUtil.ALL_POS);
+
+		for (CommerceShippingMethod commerceShippingMethod :
+				commerceShippingMethods) {
+
+			if (key.equals(commerceShippingMethod.getEngineKey())) {
+				_commerceShippingMethodLocalService.
+					deleteCommerceShippingMethod(
+						commerceShippingMethod.getCommerceShippingMethodId());
+			}
+		}
+	}
+
+	@Modified
+	protected void modified(Map<String, Object> properties)
+		throws PortalException {
+
+		_functionCommerceShippingEngineConfiguration =
+			ConfigurableUtil.createConfigurable(
+				FunctionCommerceShippingEngineConfiguration.class, properties);
+
+		List<CommerceShippingMethod> commerceShippingMethods =
+			_commerceShippingMethodLocalService.getCommerceShippingMethods(
+				QueryUtil.ALL_POS, QueryUtil.ALL_POS);
+
+		for (CommerceShippingMethod commerceShippingMethod :
+				commerceShippingMethods) {
+
+			String key = (String)properties.get("key");
+
+			if (key.equals(commerceShippingMethod.getEngineKey())) {
```

Comment:
can this key be null?

Reply:
no it can't be null, the class would not be instantiated in that case, and so we would not even reach the activate method and so the update